src/properties/forest.c

```diff
@@ -0,0 +1,215 @@
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_structural.h"
+
+#include "igraph_adjlist.h"
+#include "igraph_constructors.h"
+#include "igraph_interface.h"
+#include "igraph_stack.h"
+
+/* count the number of vertices reachable from the root */
+static int igraph_i_is_forest_visitor(igraph_integer_t root,igraph_vector_t *visited, const igraph_adjlist_t *al, igraph_integer_t *visited_count,igraph_integer_t *res,  igraph_neimode_t mode){
+
+    igraph_stack_int_t stack;
+    igraph_integer_t i;
+    IGRAPH_CHECK(igraph_stack_int_init(&stack, 0));
+    IGRAPH_FINALLY(igraph_stack_int_destroy, &stack);
+    /* push the root into the stack */
+    IGRAPH_CHECK(igraph_stack_int_push(&stack, root));
+
+    while (! igraph_stack_int_empty(&stack)) {
+        igraph_integer_t u;
+        igraph_vector_int_t *neighbors;
+        igraph_integer_t ncount;
+        u = igraph_stack_int_pop(&stack);
+
+        /* Take a vertex from stack and check if it is already visited
+         * if yes, then graph is not a forest
+         * else add it to the visited vector
+         */
+        if (IGRAPH_LIKELY(! VECTOR(*visited)[u])) {
+            VECTOR(*visited)[u] = 1;
+            *visited_count += 1;
+        }
+        else{
+            *res=0;
+            break;
+        }
+        /* register all its neighbours (except its parent) for future processing */
+        neighbors = igraph_adjlist_get(al, u);
+        ncount = igraph_vector_int_size(neighbors);
+
+        for (i = 0; i < ncount; ++i) {
+            igraph_integer_t v = VECTOR(*neighbors)[i];
+            if(mode==IGRAPH_ALL){
+                if (IGRAPH_LIKELY(!VECTOR(*visited)[v])){
+                    IGRAPH_CHECK(igraph_stack_int_push(&stack, v));
+                }
+                /*to check for loop in undirected graph*/
+                else if(v==u){
+                    *res=0;
+                    break;
+                }
+            }
+            else{
+                IGRAPH_CHECK(igraph_stack_int_push(&stack, v));
+            }
+        }
+    }
+    igraph_stack_int_destroy(&stack);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    return IGRAPH_SUCCESS;
+}
+
+/**
+ * \ingroup structural
+ * \function igraph_is_forest
+ * \brief Decides whether the graph is a forest.
+ *
+ * An undirected graph is a forest if it has no cycles.
+ * </para><para>
+ *
+ * In the directed case, a possible additional requirement is that two
+ * different trees should not have common vertex. Also, edges in each
```

Comment:
Isn't there something missing from this definition? It seems to allow for a vertex with two parents. Like 0->1, 0->2, 1->3, 2->3 should not be in forest, right? I see that it's also in `is_tree`, but I'm still confused.

Reply:
Oh thats right. That condition is checked in the code, I will add it in the description too.